### PR TITLE
RUST-1064 Retry on handshake failure

### PR DESCRIPTION
**Pull request by abr-egn**
RUST-1064

This updates operation logic to retry if a connection fails with a network or authentication error.  The test sync for this also brought a split to `legacy` and `unified`; none of the tests in `legacy` were modified.

### CODE REVIEW

**Review by kmahar — COMMENTED** [review]
it looks like from the ticket there are also new retryable writes and txns tests to sync (though we are maybe supposed to skip the txn tests for now)

**Review by patrickfreed — APPROVED** [review]
LGTM! glad to see the existing pool cleared error retry mechanism ended up being the same for the more general one

**Comment by patrickfreed** [comment]
From looking at the test failures, it looks like a load balancer test (the "errors during the initial connection hello are ignored" test in `sdam-error-handling.yml`) needs to be updated, though I'm not sure it actually ever was as part of the spec changes. We may need to file a DRIVERS ticket about that.

**Comment by kmahar** [comment]
> From looking at the test failures, it looks like a load balancer test (the "errors during the initial connection hello are ignored" test in sdam-error-handling.yml) needs to be updated, though I'm not sure it actually ever was as part of the spec changes. We may need to file a DRIVERS ticket about that.

it seems strange to me that Python, Java, and .NET are all still passing that test -- at least I'm guessing they are based on the fact they have merged in PRs and closed their tickets for this. I suggest asking in #driver-devs to try to figure out why only we have run into this so far

**Comment by abr-egn** [comment]
My theory is that I need to refine the Rust driver error-handling machinery to make a more precise distinction between error types.

File: src/client/auth/sasl.rs
Comment (abr-egn) [inline src/client/auth/sasl.rs:103]: When implementing this, I was a little surprised to learn that the SASL commands use bespoke parsing and validation rather than going through Serde integration like the rest; if there's no strong reason for that, it might be a good candidate for future cleanup.

File: src/client/executor.rs
Comment (abr-egn) [inline src/client/executor.rs:334]: This seems a little loopy but:
* the spec mandates that the `RETRYABLE_WRITE_ERROR` label be added for retryable errors from authentication and
* `is_write_retryable()` is implemented as a label check, so it can't be used to add the label here in the first place

Note that the csharp logic, which is [called out](https://github.com/mongodb/specifications/blob/082306075eb2d125c47503e94fabc42f00a16784/source/retryable-reads/retryable-reads.rst#reference-implementation) in the spec as a reference implementation, ends up doing the [same thing](https://github.com/mongodb/mongo-csharp-driver/blob/21f8c1615dcfec501b66bc19a43e185e13fdad46/src/MongoDB.Driver.Core/Core/Operations/RetryableWriteOperationExecutor.cs#L139).

File: src/client/executor.rs
Comment (abr-egn) [inline src/client/executor.rs:827]: It turns out this didn't need to be async.

**Review by abr-egn — COMMENTED** [review]
Updated, PTAL.

File: src/client/executor.rs
Comment (patrickfreed) [reply src/client/executor.rs:334]: `err.add_labels_and_update_pin` calls through to `should_add_retryable_write_label`, which _should_ take care of this for us, but it doesn't currently for two reasons:
  - we always pass in `None` for retryability
      - This can be fixed by creating a new `Retryability` variant called `Always` or something
  - more importantly, we don't have a connection and therefore don't know the max wire version of the server, so we don't know if we're supposed to add labels to it or should leave it to the server to do so
      - @matthewdale points out the same issue as experienced in the Go driver [here](https://jira.mongodb.org/browse/GODRIVER-2191?focusedCommentId=4422059&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-4422059).

I think what we ideally need here is a single set of criteria for retrying handshake errors that is independent of whether the operation is a read or write or what the server version might be. It feels like it'll end up just being network errors + the union of the pre-4.4 retryable write + retryable read error codes. but I'm not sure.

With all that being said, I think that the solution you have here seems good until the spec gets that all fleshed out.

**Review by patrickfreed — COMMENTED** [review]
The code looks good to me as-is, but I'm wondering if we should hold off for a bit pending some further spec changes that might come as a result of the issues we've faced here. I left a comment on DRIVERS-746 about it.

File: src/client/executor.rs
Comment (kmahar) [reply src/client/executor.rs:334]: should we be checking here if the client has `retryWrites` set to true before we add the label?

spec: "When the driver encounters a network error establishing an initial connection to a server,
  it MUST add a RetryableWriteError label to that error if the MongoClient performing
  the operation has the retryWrites configuration option set to true."

and I see C# checks `context.RetryRequested` there.

do any of the new tests cover retryWrites=false? (I'm not seeing any that do)

I guess either way it seems like we only actually do retry if the client has it set given the check in `get_op_retryability`, but I'm wondering if there could be any ill effects or user confusion from adding the label unconditionally here -- the retryable writes spec discusses that a bit [here](https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#why-does-the-driver-only-add-the-retryablewriteerror-label-to-errors-that-occur-on-a-mongoclient-with-retrywrites-set-to-true)

**Review by patrickfreed — APPROVED** [review]

File: src/client/executor.rs
Comment (abr-egn) [reply src/client/executor.rs]: Good catch - updated to check for `retryWrites` here.

**Review by kmahar — APPROVED** [review]

**Review by isabelatkinson — APPROVED** [review]

**Pull request merged**
